store.go

```diff
@@ -0,0 +1,170 @@
+
+	expiresAt time.Time
+}
+
+type expirableStore struct {
+	maxSize int
+
+	items map[string]*entry
+
+	buckets            []bucket
+	bucketTTL          time.Duration
+	numBuckets         int
+	nextBucketToExpire int
+
+	mu sync.Mutex
+
+	pool sync.Pool
+}
+
+func newStore(ctx context.Context, maxSize int, maxEntryTTL time.Duration, o ...Option) (store, error) {
+	const op = "rate.newStore"
+
+	opts := getOpts(o...)
+
+	switch {
+	case maxSize <= 0:
+		return nil, fmt.Errorf("%s: max size must be greater than zero: %w", op, ErrInvalidMaxSize)
+	case maxEntryTTL <= 0:
+		return nil, fmt.Errorf("%s: max entry ttl must be greater than zero: %w", op, ErrInvalidParameter)
+	case opts.withNumberBuckets <= 0:
+		return nil, fmt.Errorf("%s: number of buckets must be greater than zero: %w", op, ErrInvalidNumberBuckets)
+	}
+
+	bucketTTL := (maxEntryTTL) / time.Duration(opts.withNumberBuckets-1)
+
+	buckets := make([]bucket, opts.withNumberBuckets)
+	for i := 0; i < opts.withNumberBuckets; i++ {
+		buckets[i] = bucket{
+			entries: make(map[string]*entry, maxSize),
+		}
+	}
+
+	s := &expirableStore{
+		maxSize:    maxSize,
+		items:      make(map[string]*entry, maxSize),
+		buckets:    buckets,
+		bucketTTL:  bucketTTL,
+		numBuckets: opts.withNumberBuckets,
+		pool: sync.Pool{
+			New: func() any {
+				return &entry{
+					value: &Quota{},
+				}
+			},
+		},
+	}
+
+	go s.deleteExpired(ctx)
+	return s, nil
+}
+
+func (s *expirableStore) deleteExpired(ctx context.Context) {
+	ticker := time.NewTicker(s.bucketTTL)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
+			s.emptyExpiredBucket()
+		}
+	}
+}
+
+func (s *expirableStore) fetch(id string, limit *Limit) (*Quota, error) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	key := getKey(limit.Resource, limit.Action, string(limit.Per), id)
+
+	e, ok := s.items[key]
+	switch {
+	case !ok:
+		e = s.pool.Get().(*entry)
+		e.key = key
+		e.value.reset(limit)
+		if err := s.add(e); err != nil {
+			s.pool.Put(e)
+			return nil, err
+		}
+		s.addToBucket(e)
+	case e.value.Expired():
+		s.removeFromBucket(e)
+		e.value.reset(limit)
+		s.addToBucket(e)
+	}
+
+	return e.value, nil
+}
+
+func (s *expirableStore) add(e *entry) error {
+	if _, ok := s.items[e.key]; !ok && len(s.items) >= s.maxSize {
+		return ErrLimiterFull
+	}
+	s.items[e.key] = e
+	return nil
+}
+
+func (s *expirableStore) addToBucket(e *entry) {
+	e.bucket = (int(e.value.limit.Period/s.bucketTTL) + s.nextBucketToExpire) % s.numBuckets
+	s.buckets[e.bucket].entries[e.key] = e
+	if s.buckets[e.bucket].expiresAt.Before(e.value.expiresAt) {
+		s.buckets[e.bucket].expiresAt = e.value.expiresAt
+	}
+}
+
+func (s *expirableStore) emptyExpiredBucket() {
+	s.mu.Lock()
+
+	toExpire := s.nextBucketToExpire
+	s.nextBucketToExpire = (s.nextBucketToExpire + 1) % s.numBuckets
+
+	timeToExpire := time.Until(s.buckets[toExpire].expiresAt)
+	if timeToExpire > 0 {
+		s.mu.Unlock()
+		time.Sleep(timeToExpire)
```

Comment:
can you add a comment about the expectation for `emptyExpiredBucket()` to be
called in a per bucket goroutine

Reply:
Added some comments to this method. I also added some TODO for some additional documentation around the expireableStore that I will address in a future PR along with other documentation TODOs.